### ISSUE TITLE
/stream download a file

### ISSUE THREAD

**Issue by darwinjob** [irods/irods_client_rest_cpp#117]
Simplest possible use case: I need to download a file, I have a token or a ticket.
So I use /stream endpoint. But according to the documentation I need to provide the count parameter. Naturally I don’t know the file size. What’s the solution for this?

**Comment by korydraughn** [comment]
The endpoint is modeled after POSIX read/write.

The endpoint is meant to be called multiple times with different offsets. You never want to pass the full size of the object via the `count` parameter. The `count` parameter will always result in the REST server allocating a buffer of `count` bytes in size on each request.

Here is a little bit of pseudo code to demonstrate the idea of how to read a file of any size:
```cpp
int fd = // Open the data object.

// "count" should always be set to a reasonable value to avoid
// out-of-memory errors in the server.
int count = 1024 * 1024 * 2;
char buf[count];
int offset = 0;

// Do an incremental read of the data object.
while (true) {
    int bytes_read = read(fd, buf, count, offset);

    if (bytes_read < 0) {
        handle_error(bytes_read);
        break;
    }

    offset = offset + bytes_read;
    
    if (bytes_read < count) {
        do_something_with_buffer(buf, bytes_read);

        // We received fewer bytes than requested.
        // We've read all of the bytes.
        // Exit the loop.
        break;
    }

    do_something_with_buffer(buf, bytes_read);
}
```

Hope that helps :)

**Comment by darwinjob** [comment]
Yes, this helps, Kory, thank you :)
Let me rephrase this a little.

I have a file stored in iRODS. I would like to share that file via URL. Meaning I send that URL/link to my colleague. He/she should be able to download that file simply by clicking on the link. How would you do this? No Metalnx please :)

**Comment by korydraughn** [comment]
There is currently no way to do that without some additional code to drive the download.

Something has to make multiple calls to incrementally download the data object. Now, if you have a very large machine with seemingly infinite memory, then maybe querying for the data size and passing that as the `count` argument would be acceptable.

You could write a Python client or web app that drives the REST API too.

**Comment by darwinjob** [comment]
This is what I was afraid of 😐
Thank you, Kory 🙂
One more question if you don't mind. Is it possible to imagine /stream URL without token or ticket?

**Comment by trel** [comment]
This will not be implemented in this repository.

Please see the new HTTP API:

https://github.com/irods/irods_client_http_api
